### ISSUE TITLE
Windows: ICEs and blockers when building a VST plugin

### ISSUE THREAD

**Issue by p0nce** [ldc-developers/ldc#943]
**Problem**

Trying to build medium-sized shared libraries for Windows fail in various ways, among which ICEs.
Specifically this project: https://github.com/p0nce/dplug/tree/master/examples/distort

**Setup:**
Windows 8 64-bit
- For 32-bit builds: ldc2-0.15.2-beta1-mingw-x86 for 32-bit builds with latest mingw-w64 toolchain
- For 64-bit builds: ldc2-0.15.2-beta1-win64-msvc for 64-bit builds with MSVC 12.0 linker

**To reproduce:**
- checkout dplug: https://github.com/p0nce/dplug

- go to the distort example: https://github.com/p0nce/dplug/tree/master/examples/distort

- Build for 64-bit with DUB: `dub build --compiler=ldc --arch=x86_64`
(use latest DUB to avoid bad 'x86_64' flags)

This will ask for many .lib that can be found in the Windows SDK. 
```
Assertion failed: !(Pair.first != Ty && Pair.second == Ty) && "mapping to a sour
ce type", file D:\Temp\LDC-BUILD\src\llvm-3.6.0.src\lib\Linker\LinkModules.cpp,
line 265
```

**Another blocker that puzzles me**

- Build a 32-bit debug version: `dub build --compiler=ldc --arch=x86 --combined`
```
D:/msys/build/src/ldc/runtime/druntime/src/core/sys/windows/dll.d:386: undefined
 reference to `_tlsstart'
D:/msys/build/src/ldc/runtime/druntime/src/core/sys/windows/dll.d:386: undefined
 reference to `_tlsend'
D:/msys/build/src/ldc/runtime/druntime/src/core/sys/windows/dll.d:386: undefined
 reference to `_tls_callbacks_a'
```

**Comment by kinke** [comment]
Windows is work in progress. The imo best option is to use a brand-new LDC based on the merge-2.067 branch, with LLVM master. That will pass almost all druntime & phobos unit tests. See http://forum.dlang.org/post/cejrytcqfooirbhadmsa@forum.dlang.org.

**Comment by redstar** [comment]
The merge-2.067 is really recommended because the beta1 versions you are using has some serious bugs, e.g. ldmd not working at all.

**Comment by p0nce** [comment]
OK, I'll wait more then.

**Comment by redstar** [comment]
Building ldc on Windows is not really hard, see http://wiki.dlang.org/Building_and_hacking_LDC_on_Windows_using_MSVC.

**Comment by p0nce** [comment]
This problem is gone with LDC 0.15.2 beta2:

Assertion failed: !(Pair.first != Ty && Pair.second == Ty) && "mapping to a sour
ce type", file D:\Temp\LDC-BUILD\src\llvm-3.6.0.src\lib\Linker\LinkModules.cpp,
line 265

Now both 32-bit and 64-bit versions give:
```
c:\d\ldc-32b\bin/../lib/libdruntime-ldc.a(dll.obj):(.text[__D4core3sys7windows3d
ll18dll_process_attachFPvbZb]+0x12): undefined reference to `_tlsstart'
c:\d\ldc-32b\bin/../lib/libdruntime-ldc.a(dll.obj):(.text[__D4core3sys7windows3d
ll18dll_process_attachFPvbZb]+0x1a): undefined reference to `_tlsend'
c:\d\ldc-32b\bin/../lib/libdruntime-ldc.a(dll.obj):(.text[__D4core3sys7windows3d
ll18dll_process_attachFPvbZb]+0x21): undefined reference to `_tls_callbacks_a'
collect2.exe: error: ld returned 1 exit status
```
Back again to Windows DMD for now...

**Comment by redstar** [comment]
The assertion is fixed in LDC 0.16.0.
